lib/9pfs/9pfs_vnops.c

```diff
@@ -1016,10 +1016,92 @@ static int uk_9pfs_ioctl(struct vnode *dvp, struct vfscore_file *fp,
 			 unsigned long com, void *data)
 {
 	switch (com) {
+	/**
+	 * HACK: In binary compatibility mode, Ruby tries to set O_ASYNC,
+	 * which Unikraft does not yet support. If the `ioctl` call returns
+	 * an error, Ruby stops working, even if it does not depend on the
+	 * O_ASYNC being properly set.
+	 *
+	 * Until proper support, just return 0 in case an `FIONBIO` ioctl
+	 * request is done, and ENOTSUP for all other cases.
+	 *
+	 * Setting `ioctl` to a nullop will not work, since it is used by
+	 * interpreted languages (e.g. python3) to check if it should start
+	 * the interpretor or just read a file.
+	 */
+	case FIONBIO:
+		return 0;
+	/* For every `ioctl` request related to a terminal, return ENOTTY. */
+	case TCGETS:
+	case TCSETS:
+	case TCSETSW:
+	case TCSETSF:
+	case TCGETA:
+	case TCSETA:
+	case TCSETAW:
+	case TCSETAF:
+	case TCSBRK:
+	case TCXONC:
+	case TCFLSH:
+	case TIOCEXCL:
+	case TIOCNXCL:
+	case TIOCSCTTY:
+	case TIOCGPGRP:
+	case TIOCSPGRP:
+	case TIOCOUTQ:
+	case TIOCSTI:
 	case TIOCGWINSZ:
+	case TIOCSWINSZ:
+	case TIOCMGET:
+	case TIOCMBIS:
+	case TIOCMBIC:
+	case TIOCMSET:
+	case TIOCGSOFTCAR:
+	case TIOCSSOFTCAR:
+	case FIONREAD:
+	case TIOCLINUX:
+	case TIOCCONS:
+	case TIOCGSERIAL:
+	case TIOCSSERIAL:
+	case TIOCPKT:
+	case TIOCNOTTY:
+	case TIOCSETD:
+	case TIOCGETD:
+	case TCSBRKP:
+	case TIOCSBRK:
+	case TIOCCBRK:
+	case TIOCGSID:
+	case TIOCGRS485:
+	case TIOCSRS485:
+	case TIOCGPTN:
+	case TIOCSPTLCK:
+	case TIOCGDEV:
+	case TCGETX:
+	case TCSETX:
+	case TCSETXF:
+	case TCSETXW:
```

Comment:
Are you still sure these are all? WHat about RS485 related configs? (`TIOCGRS485` or `TIOCSRS485`). 
Also, notice this is looking huge. `ioctl` numbers usually follow a pattern depending on the family they appertain to. Check out these macro's:
```C
#define _IO(a,b) _IOC(_IOC_NONE,(a),(b),0)
#define _IOW(a,b,c) _IOC(_IOC_WRITE,(a),(b),sizeof(c))
#define _IOR(a,b,c) _IOC(_IOC_READ,(a),(b),sizeof(c))
#define _IOWR(a,b,c) _IOC(_IOC_READ|_IOC_WRITE,(a),(b),sizeof(c))
```
Also take a look at Linux's `include/uapi/asm-generic/ioctls.h`. Notice that all `tty` related ioctl requests have either type `t` or type `T` (Take note of this comment in the file `/* 0x54 is just a magic number to make these relatively unique ('T') */`). Could we, maybe, rewrite this massive switch case to just use something like this macro instead?
```C
#define IS_TTY_REQ(req)                                   (req & ('T' << 8))
```
and then
```C
if (com == FIONBIO)
    return 0;
else if  (IS_TTY_REQ(com)):
                return ENOTTY;
 
 return ENOTSUP;
```

Reply:
Yeah that's really nice. I looked for a pattern in the tty reqs but I don't know how I missed that, thanks.

I put the `IS_TTY_REQ` macro in the `9p_vnops.c` file, but it might be better to have it available in a header somewhere (`vfscore/include/something`?), I just did not find the proper place yet, so if you have any ideas let me know.

Reply:
Good point :+1:.  Probably should be in a header accessible to vfscore drivers only, so maybe `lib/vfscore/include/vfscore/vnode.h`? And have it renamed to something like `IOCTL_REQ_IS_TTY` :thinking:? Let's see what @skuenzer has to say about this, he will probably have a better idea.

Reply:
Yes, `vfscore/vnode.h` makes sense. I added it there for now, will wait for @skuenzer's input.